text/0000-export.md

```diff
@@ -0,0 +1,401 @@
+- There are now two library crates: one that will be compiled into the dynamic library (the .dll/.so/.dylib file),
+  and one that provides the bindings to that dynamic library.
+  The second library likely fully inlined into the final application,
+  as it only has wrappers, just to bring back the original (safe) function signatures.
+
+Much of this solution could be automated by a procedural macro,
+but splitting a library crate in two falls outside of what a procedural macro can reasonably do.
+
+### Proposed solution sketch
+
+Instead of all the manual usage of `#[no_mangle]`, `extern`, and `unsafe`,
+a much better solution would look as closely as possible to the original code.
+
+With the proposal below, one only needs to add an `#[export]` attribute, and give the function a stable ABI
+(e.g. `extern "C"` or (in the future) `extern "crabi"`):
+
+```rust
+//! library crate
+
+#[export]
+pub extern "C" fn hello() {
+    println!("Hello!");
+}
+```
+
+```rust
+//! application crate
+
+fn main() {
+    library::hello();
+}
+```
+
+The library can then be either linked statically or dynamically, by informing cargo of the choice:
+
+```diff
+  [dependencies]
+- library = { path = "..." }
++ library = { path = "...", dynamic = true }
+```
+
+## Proposal
+
+Creating and using dynamic libraries involves three things:
+
+1. A stable ABI that can be used for the items that are exported/imported.
+2. A way to export and import items.
+3. A way to create and use dynamic libraries.
+
+For (1) we currently only have `extern "C"`, which only suffices for very simple cases.
+This proposal does not include any improvements for (1),
+but the ["crabi" proposal](https://github.com/rust-lang/rust/pull/105586) proposes the creation
+of a new `extern "…"` ABI that is more flexible, which perfectly complements this proposal.
+
+This proposal provides solutions for (2) and (3).
+Exporting (and importing) items is done through a new language feature: the `#[export]` attribute.
+Creating and using dynamic libraries is made easy through a new Cargo feature: `dynamic` dependencies.
+
+### The `#[export]` Attribute
+
+The `#[export]` attribute is used to mark items which are "stable" (in ABI/layout/signature)
+such that they can be used across the border between (separately compiled) dynamically linked libraries/binaries.
+
+The `#[export]` attribute can be applied to any public item that is *exportable*.
+Which items are *exportable* is something that can increase over time with future proposals.
+Initially, only the following items are *exportable*:
+
+- Non-generic functions with a stable ABI (e.g. `extern "C"`)
+  for which every user defined type used in the signature is also marked as `#[export]`.
+  - This includes type associated functions ("methods").
+- Structs/enums/unions with a stable representation (e.g. `repr(i32)` or `repr(C)`).
+- Re-exports of those items (`use` statements, `type` aliases).
```

Comment:
Two questions about edge cases here:
- For `#[export] pub struct Foo;` is the `Foo` _value_ (not the _type_) also exported? My read based on this is "no" and that might be a surprising difference between using a Rust crate as a static vs dynamic dependency.
- For `#[export] pub struct Foo();` is the implicit Ctor also exported? My expectation based on this is "yes" but it would be great to clarify explicitly.

These are on my mind because they were a problem recently in https://github.com/rust-lang/rust/issues/111338 and https://predr.ag/blog/breaking-semver-in-rust-by-adding-private-type-or-import/

Reply:
Great question. I don't have an answer to that now. I'll think about it a bit. :)

Reply:
I think neither of those examples should result in any symbols being exported from the dynamic library, but I think in both cases the value/function should be usable in a crate that imports the library.

Dynamically linking the `Foo()` function or `Foo` value does not make sense, since they only have one possible trivial definition based on the type definition. (It's not possible for the implementation of the `Foo()` function to change as long as the type is just `struct Foo();`.)

(Perhaps we can simply think of this as a case of the 'inline export' concept described in https://github.com/rust-lang/rfcs/pull/3435#discussion_r1202571220)

Reply:
Would the answer change for `#[export] pub struct Foo(pub Bar, pub Baz);` at all, assuming `Bar` and `Baz` are exported types?

I don't have an intuitive understanding of inline exports yet (specifically how they are different than normal static linking _for those items specifically_), so I wanted to clarify the boundaries on "trivial definition" — whether it's only for empty tuples or also for more complex tuple structs where things like alignment and padding might also be in play.

Reply:
Same answer for that case. The constructor of `Foo` does not need to be dynamically linked, if the only possible implemementation is just to fill in the fields which are already public and promised to be ABI stable.

`Foo { 0: .., 1: .. }` and `Foo(.., ..)` should just do the same. Neither should result in a function being called in a dynamic library.